### PR TITLE
Fix base64 request decoding for grpc-web-text

### PR DESCRIPTION
**Pull request by drmontgomery**
Added DecodeBase64 flow that enforces 4-byte alignment.
Added preDecodeFlow parameter to AbstractGrpcProtocol.
Added tests.

References #1650

### CODE REVIEW

**Comment by drmontgomery** [comment]
Re binary compatibility: 
```
[error]  * abstract method preDecode(akka.util.ByteString)akka.util.ByteString in class akka.grpc.internal.GrpcProtocolWebBase does not have a correspondent in current version
[error]  * abstract method preDecodeFlow()akka.stream.scaladsl.Flow in class akka.grpc.internal.GrpcProtocolWebBase is present only in current version
[error]  * abstract method preDecodeStrict(akka.util.ByteString)akka.util.ByteString in class akka.grpc.internal.GrpcProtocolWebBase is present only in current version
[error]  * method preDecode(akka.util.ByteString)akka.util.ByteString in object akka.grpc.internal.GrpcProtocolWeb does not have a correspondent in current version
[error]  * method preDecode(akka.util.ByteString)akka.util.ByteString in object akka.grpc.internal.GrpcProtocolWebText does not have a correspondent in current version
[error]  * method reader(akka.grpc.internal.Codec,scala.Function2,scala.Function1)akka.grpc.GrpcProtocol#GrpcProtocolReader in object akka.grpc.internal.AbstractGrpcProtocol does not have a correspondent in current version
[error]  * static method preDecode(akka.util.ByteString)akka.util.ByteString in class akka.grpc.internal.GrpcProtocolWeb does not have a correspondent in current version
[error]  * static method preDecode(akka.util.ByteString)akka.util.ByteString in class akka.grpc.internal.GrpcProtocolWebText does not have a correspondent in current version
```
I think it's safe to filter all of these, since they're entirely contained within akka.grpc.internal (both definitions and potential references).

**Comment by drmontgomery** [comment]
Re DecodeBase64:
The DecodeBase64 flow is effectively doing two things: enforcing the 4-byte chunk alignment and then passing the aligned chunks into ByteString.decodeBase64. It feels like the chunk alignment aspect could be useful and reusable on its own, perhaps as an addition to akka.stream.scaladsl.Framing.

Example:
```
def decodeBase64: Flow[ByteString, ByteString, NotUsed] = 
    Flow[ByteString]
        .via(Framing.aligned(4))
        .map(_.decodeBase64)
```

Is this worth pursuing, either for this PR or as a PR for akka streams?

File: runtime/src/main/scala/akka/grpc/internal/DecodeBase64.scala [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+          buffer = remaining
+        } else {
+          pull(in)
+        }
+      }
+
+      override def onUpstreamFinish(): Unit = {
+        if (buffer.nonEmpty) {
+          if (isAvailable(out)) {
+            push(out, buffer.decodeBase64)
```
Comment (patriknw) [inline runtime/src/main/scala/akka/grpc/internal/DecodeBase64.scala:55]: Can we be sure that the right number of bytes have been pushed to the buffer, or is there a risk of some premature upstream finish? Maybe best to have the same `val decodeLength = length - length % 4` logic here?

Comment (patriknw) [reply runtime/src/main/scala/akka/grpc/internal/DecodeBase64.scala:55]: and same in the push from onPull

**Review by patriknw — COMMENTED** [review]
looking good

File: runtime/src/main/scala/akka/grpc/internal/DecodeBase64.scala (hunk 1 above)
Comment (drmontgomery) [reply runtime/src/main/scala/akka/grpc/internal/DecodeBase64.scala:55]: Padding in base64 is only required for correctness when concatenating encoded chunks. Most decoders will implicitly add padding for any missing bytes at the end of encoded data. I think this is an instance of the robustness principle ("be conservative in what you send, liberal in what you accept").

akka.util.ByteString.decodeBase64 automatically adds trailing padding:
```
scala> val data = akka.util.ByteString(List(0.toByte))
val data: akka.util.ByteString = ByteString(0)

scala> val encoded = data.encodeBase64
val encoded: akka.util.ByteString = ByteString(65, 65, 61, 61)

scala> val decoded = encoded.dropRight(2).decodeBase64
val decoded: akka.util.ByteString = ByteString(0)
```

I followed this same principle when implementing the DecodeBase64 flow. If the stream ends with an unaligned chunk of data, we decode it and pass it downstream before completing the stage. This should allow the akka-grpc servers to work with grpc-web client implementations that omit padding, without affecting correctness.

Comment (patriknw) [reply runtime/src/main/scala/akka/grpc/internal/DecodeBase64.scala:55]: I'm not sure we talk about the same problem. Here in onUpstreamFinish it tries to `decodeBase64` whatever is currently in the buffer. For example, `ByteString.fromString("a")` but only 1 byte has been pushed to the buffer and then the stream is finished. That can probably not happen for well behaved encoding upstreams but should this flow fail in such case or should it ignore the incomplete trailing part?

```
scala> ByteString.fromString("a").encodeBase64.take(1).decodeBase64.toString
java.lang.IllegalArgumentException: Input byte[] should at least have 2 bytes for base64 bytes
  at java.base/java.util.Base64$Decoder.decodedOutLength(Base64.java:709)
  at java.base/java.util.Base64$Decoder.decode(Base64.java:662)
  at akka.util.ByteString$ByteString1.decodeBase64(ByteString.scala:384)
```

Comment (raboof) [reply runtime/src/main/scala/akka/grpc/internal/DecodeBase64.scala:55]: > That can probably not happen for well behaved encoding upstreams but should this flow fail in such case or should it ignore the incomplete trailing part?

I don't think we should silently ignore/drop unaligned input.

For unaligned-but-parsable input (`ByteString.fromString("a").encodeBase64.take(2)`) I'm OK with either parsing it or throwing an error. The current implementation parses it, which seems fine by me.

For unaligned unparsable input (`ByteString.fromString("a").encodeBase64.take(1)`) we should fail the stream. Here we could either explicitly `fail`/`failStage` or just allow the exception to be thrown and have akka streams deal with it. I'm not sure I have a strong preference between those two, you?

Comment (patriknw) [reply runtime/src/main/scala/akka/grpc/internal/DecodeBase64.scala:55]: ok, sounds good to me. Then current code does the right thing already.

**Review by raboof — APPROVED** [review]

**Pull request merged**